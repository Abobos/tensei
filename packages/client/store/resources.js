import React from 'react'

const Resources = React.createContext({
    resources: []
})

export const withResources = Component => {
    const WithResourcesComponent = props => {
        return (
            <Resources.Consumer>
                {value => (
                    <Component
                        resources={value.resources}
                        dashboards={value.dashboards}
                        {...props}
                    />
                )}
            </Resources.Consumer>
        )
    }

    return WithResourcesComponent
}

export default Resources
