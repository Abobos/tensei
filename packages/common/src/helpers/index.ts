import { Request } from 'express'
import { ResourceContract } from '@tensei/common'

export class ResourceHelpers {
    public resource: ResourceContract | null = null

    constructor(public resources: ResourceContract[]) {}

    public findResource = (resourceSlug: string | ResourceContract) => {
        if (!resourceSlug) {
            throw {
                message: `Resource ${resourceSlug} not found.`,
                status: 404
            }
        }

        if (typeof resourceSlug !== 'string') {
            return resourceSlug
        }

        const resource = this.resources.find(
            resource =>
                resource.data.slug === resourceSlug ||
                resource.data.name === resourceSlug
        )

        if (!resource) {
            throw {
                message: `Resource ${resourceSlug} not found.`,
                status: 404
            }
        }

        return resource
    }

    getFieldFromResource = (
        resource: ResourceContract,
        databaseField: string
    ) => {
        return resource.data.fields.find(
            field =>
                field.name === databaseField ||
                field.databaseField === databaseField
        )
    }

    setResource = (resourceOrSlug: ResourceContract | string) => {
        this.resource = this.findResource(resourceOrSlug)

        return this
    }

    protected getCurrentResource = () => {
        if (!this.resource) {
            throw {
                message: `Missing resource for this operation.`,
                status: 400
            }
        }

        return this.resource
    }
}

/**
 * Apply typescript mixins to a class constructor.
 *
 * @param derivedCtor any
 * @param constructors any
 */
export const mix = (derivedCtor: any, constructors: any[]) => {
    constructors.forEach(baseCtor => {
        Object.getOwnPropertyNames(baseCtor.prototype).forEach(name => {
            const propertyDescription = Object.getOwnPropertyDescriptor(
                baseCtor.prototype,
                name
            )

            if (!propertyDescription) {
                return
            }

            Object.defineProperty(
                derivedCtor.prototype,
                name,
                propertyDescription
            )
        })
    })
}
