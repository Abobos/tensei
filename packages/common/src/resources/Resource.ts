import { id } from '../fields/ID'
import Field, { SerializedField } from '../fields/Field'
import { HookFunction, AuthorizeFunction, Permission } from '../config'

import Pluralize from 'pluralize'
import { snakeCase, paramCase, camelCase } from 'change-case'
import Action, { SerializedAction } from '../actions/Action'

interface ValidationMessages {
    [key: string]: string
}

interface ResourceData {
    name: string
    table: string
    group: string
    slug: string
    label: string
    groupSlug: string
    valueField: string
    camelCaseName: string
    displayField: string
    noTimeStamps: boolean
    perPageOptions: number[]
    permissions: Permission[]
    displayInNavigation: boolean
    validationMessages: ValidationMessages
}

interface ResourceDataWithFields extends ResourceData {
    fields: Field[]
    actions: Action[]
}

export interface SerializedResource extends ResourceData {
    fields: SerializedField[]
    actions: SerializedAction[]
}

export class Resource<ResourceType = {}> {
    private $model: ResourceType | null = null

    private $models: ResourceType[] = []

    public authorizeCallbacks: {
        authorizedToSee: AuthorizeFunction
        authorizedToCreate: AuthorizeFunction
        authorizedToUpdate: AuthorizeFunction
        authorizedToDelete: AuthorizeFunction
    } = {
        authorizedToSee: (request) => true,
        authorizedToCreate: (request) => true,
        authorizedToUpdate: (request) => true,
        authorizedToDelete: (request) => true,
    }

    public hooks: {
        beforeCreate: HookFunction
        beforeUpdate: HookFunction
        afterCreate: HookFunction
        afterUpdate: HookFunction
    } = {
        beforeCreate: (payload, request) => {
            return payload
        },

        beforeUpdate: (payload, request) => {
            return payload
        },

        afterCreate: (payload, request) => {
            return payload
        },

        afterUpdate: (payload, request) => {
            return payload
        },
    }

    constructor(name: string, tableName?: string) {
        this.setValue('name', name)
        this.setValue('slug', Pluralize(paramCase(name)))
        this.setValue('label', Pluralize(name))
        this.setValue('camelCaseName', camelCase(name))
        this.setValue('table', tableName || Pluralize(snakeCase(name)))
    }

    public Model = (): any => {
        return null
    }

    public data: ResourceDataWithFields = {
        fields: [],
        actions: [],
        table: '',
        name: '',
        slug: '',
        label: '',
        permissions: [],
        group: 'Resources',
        groupSlug: 'resources',
        displayField: 'id',
        valueField: 'id',
        noTimeStamps: false,
        camelCaseName: '',
        validationMessages: {
            required: 'The {{ field }} is required.',
            email: 'The {{ field }} must be a valid email address.',
        },
        displayInNavigation: true,
        perPageOptions: [10, 25, 50],
    }

    public permissions(permissions: Permission[]) {
        this.setValue('permissions', permissions)

        return this
    }

    public canSee(authorizeFunction: AuthorizeFunction) {
        this.authorizeCallbacks['authorizedToSee'] = authorizeFunction

        return this
    }

    public canCreate(authorizeFunction: AuthorizeFunction) {
        this.authorizeCallbacks['authorizedToCreate'] = authorizeFunction

        return this
    }

    public canUpdate(authorizeFunction: AuthorizeFunction) {
        this.authorizeCallbacks['authorizedToUpdate'] = authorizeFunction

        return this
    }

    public canDelete(authorizeFunction: AuthorizeFunction) {
        this.authorizeCallbacks['authorizedToDelete'] = authorizeFunction

        return this
    }

    public displayField(displayField: string) {
        this.setValue('displayField', displayField)

        return this
    }

    public model($model: ResourceType) {
        this.$model = $model

        return this
    }

    public models($models: ResourceType[]) {
        this.$models = $models

        return this
    }

    public fields(fields: Field[]) {
        this.setValue('fields', [
            // We'll set the primary key here. We're not ready to allow custom primary keys yet.
            id('ID'),
            ...fields,
        ])

        return this
    }

    public actions(actions: Action[]) {
        this.setValue('actions', actions)

        return this
    }

    public noTimeStamps() {
        this.setValue('noTimeStamps', true)

        return this
    }

    private setValue(key: keyof ResourceDataWithFields, value: any) {
        this.data = {
            ...this.data,
            [key]: value,
        }
    }

    public perPageOptions(perPageOptions: number[]) {
        this.setValue('perPageOptions', perPageOptions)

        return this
    }

    public displayInNavigation() {
        this.setValue('displayInNavigation', true)

        return this
    }

    public hideFromNavigation() {
        this.setValue('displayInNavigation', false)

        return this
    }

    public validationMessages(validationMessages: ValidationMessages) {
        this.setValue('validationMessages', validationMessages)

        return this
    }

    public group(groupName: string) {
        this.setValue('group', groupName)
        this.setValue('groupSlug', paramCase(groupName))

        return this
    }

    public slug(slug: string) {
        this.setValue('slug', slug)

        return this
    }

    public label(label: string) {
        this.setValue('label', label)

        return this
    }

    public serialize(): SerializedResource {
        return {
            ...this.data,
            fields: this.data.fields.map((field) => field.serialize()),
            actions: this.data.actions.map((action) => action.serialize()),
        }
    }

    public beforeCreate(hook: HookFunction) {
        this.hooks = {
            ...this.hooks,
            beforeCreate: hook,
        }

        return this
    }

    public beforeUpdate(hook: HookFunction) {
        this.hooks = {
            ...this.hooks,
            beforeUpdate: hook,
        }

        return this
    }

    public afterCreate(hook: HookFunction) {
        this.hooks = {
            ...this.hooks,
            afterCreate: hook,
        }

        return this
    }

    public afterUpdate(hook: HookFunction) {
        this.hooks = {
            ...this.hooks,
            afterUpdate: hook,
        }

        return this
    }
}

export const resource = (name: string, tableName?: string) =>
    new Resource(name, tableName)

export default Resource
