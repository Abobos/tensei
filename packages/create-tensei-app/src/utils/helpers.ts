import Fs from 'fs'
import Ora from 'ora'
import Path from 'path'
import Execa from 'execa'
import Edge from 'edge.js'
import latestVersion from 'latest-version'
import signale from './signale'

export const getAvailableTemplates = () => ['quickstart']

export const getAvailableDatabases = () => ['sqlite', 'mysql', 'postgresql']

export const getPackageVersions = (database: string, api: string) =>
    Promise.all([
        latestVersion(`@mikro-orm/${database}`),
        latestVersion(`@tensei/core`),
        latestVersion(`@tensei/auth`),
        latestVersion(`@tensei/${api}`),
        latestVersion(`nodemon`)
    ])

export const parseTemplateFiles = (
    template: string,
    templateVariables: any
) => {
    const basePath = Path.resolve(__dirname, '..', '..', 'templates', template)

    return Fs.readdirSync(basePath).map(file => [
        file,
        Edge.renderString(
            Fs.readFileSync(Path.resolve(basePath, file)).toString(),
            templateVariables
        )
    ])
}

export const projectDirectoryExists = (project: string) =>
    Fs.existsSync(Path.resolve(process.cwd(), project))

export const createProjectFolder = (project: string) =>
    Fs.mkdirSync(Path.resolve(process.cwd(), project))

export const createProjectFile = (
    project: string,
    file: string,
    content: string
) =>
    Fs.writeFileSync(
        Path.resolve(
            process.cwd(),
            project,
            file.endsWith('.edge') ? file.substring(0, file.length - 5) : file
        ),
        content
    )

export const hasYarnInstalled = () => {
    try {
        return Execa.commandSync('yarn --version').exitCode === 0
    } catch (e) {
        return false
    }
}

export const installProjectDependencies = (
    packageManager: string,
    project: string,
    runDev = true
) => {
    try {
        const spinner = Ora(
            `Installing dependencies with ${packageManager}`
        ).start()

        Execa.sync(packageManager, ['install'], {
            cwd: Path.resolve(process.cwd(), project),
            stdin: 'ignore'
        })

        spinner.stop()

        signale.success(
            `ðŸš€ Done installing project.\n\nðŸŽ‰ Congratulations on the start of something amazing!\n\nðŸ’» Run the following command to get started:\n\ncd ${project} && ${packageManager} run dev\n\n`
        )

        process.exit(0)
    } catch (e) {
        signale.error(e)

        process.exit(1)
    }
}
