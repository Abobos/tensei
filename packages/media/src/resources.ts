import {
    text,
    json,
    integer,
    hasMany,
    resource,
    belongsTo,
    bigInteger
} from '@tensei/common'

export const mediaResource = () =>
    resource('File')
        .fields([
            bigInteger('Size')
                .description('The file size in Kb')
                .rules('required', 'number')
                .searchable(),
            integer('Width').nullable(),
            integer('Height').nullable(),
            text('Original Filename').nullable().searchable(),
            text('Extension')
                .description('The file extension, for example psd, pdf, png')
                .searchable(),
            text('Mime Type').nullable().searchable(),
            text('Hash').searchable(),
            text('Path').nullable().searchable(),
            text('Alt Text').nullable(),
            text('Disk').nullable(),
            json('Disk Meta').nullable(),
            hasMany('File', 'transformations'),
            belongsTo('File').nullable()
        ])
        .hideFromNavigation()
        .hideOnInsertApi()
        .hideOnUpdateApi()
