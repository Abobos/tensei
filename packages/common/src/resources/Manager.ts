import { Request } from 'express'
import { validateAll } from 'indicative/validator'
import { resource as createResourceFn } from './Resource'
import { DataPayload, FetchAllRequestQuery } from '@tensei/common'
import {
    ActionResponse,
    DatabaseRepositoryInterface,
    ResourceContract,
    ManagerContract,
} from '@tensei/common'

export class Manager implements ManagerContract {
    constructor(
        private resources: ResourceContract[],
        private db: DatabaseRepositoryInterface
    ) {}

    public findResource = (resourceSlug: string | ResourceContract) => {
        if (!resourceSlug) {
            throw {
                message: `Resource ${resourceSlug} not found.`,
                status: 404,
            }
        }

        if (typeof resourceSlug !== 'string') {
            return resourceSlug
        }

        const resource = this.resources.find(
            (resource) => resource.data.slug === resourceSlug
        )

        if (!resource) {
            throw {
                message: `Resource ${resourceSlug} not found.`,
                status: 404,
            }
        }

        return resource
    }

    public async deleteById(
        request: Request,
        resourceSlugOrResource: string | ResourceContract,
        id: number | string
    ) {
        const resource = this.findResource(resourceSlugOrResource)

        await this.db.deleteById(resource, id)
    }

    public async createAdmin(
        request: Request,
        resourceSlugOrResource: string | ResourceContract,
        payload: DataPayload
    ) {
        const resource = this.findResource(resourceSlugOrResource)

        const roleResource = this.resources.find(
            (resource) => resource.data.slug === 'administrator-roles'
        )

        if (!roleResource) {
            throw {
                message: `The role resource must be registered.`,
                status: 422,
            }
        }

        const superAdmin = await this.db.findOneByField(
            roleResource,
            'slug',
            'super-admin'
        )

        if (!superAdmin) {
            throw {
                message: `The super-admin role must be setup before creating an administrator user.`,
                status: 422,
            }
        }

        const { id } = await this.create(request, resource, {
            ...payload,
            administrator_roles: [superAdmin.id],
        })

        return id
    }

    public async create(
        request: Request,
        resourceSlugOrResource: string | ResourceContract,
        payload: DataPayload
    ) {
        const resource = this.findResource(resourceSlugOrResource)

        let validatedPayload = await this.validate(payload, resource)

        const {
            nonRelationshipFieldsPayload,
            relationshipFieldsPayload,
        } = this.breakFieldsIntoRelationshipsAndNonRelationships(
            await resource.hooks.beforeCreate(
                {
                    ...validatedPayload.parsedPayload,
                    ...validatedPayload.relationshipFieldsPayload,
                },
                request
            ),
            resource
        )

        // TODO: Insert beforeCreate hook for fields here.

        const model = await this.db.create(
            resource,
            nonRelationshipFieldsPayload,
            relationshipFieldsPayload
        )

        return model
    }

    public async update(
        request: Request,
        resourceSlugOrResource: string | ResourceContract,
        id: number | string,
        payload: DataPayload,
        patch = true
    ) {
        const resource = this.findResource(resourceSlugOrResource)

        let { parsedPayload, relationshipFieldsPayload } = await this.validate(
            payload,
            resource,
            false,
            id
        )

        parsedPayload = resource.hooks.beforeUpdate(parsedPayload, request)

        // TODO: Add beforeUpdate hook for fields.

        return this.db.update(
            resource,
            id,
            parsedPayload,
            relationshipFieldsPayload,
            patch
        )
    }

    public async updateRelationshipFields(
        resource: ResourceContract,
        payload: DataPayload,
        modelId: string | number
    ) {
        const {
            relationshipFieldsPayload,
        } = this.breakFieldsIntoRelationshipsAndNonRelationships(
            this.getResourceFieldsFromPayload(payload, resource),
            resource
        )

        const relationshipFields = resource
            .serialize()
            .fields.filter((field) => field.isRelationshipField)

        for (let index = 0; index < relationshipFields.length; index++) {
            const field = relationshipFields[index]
            const relatedResource = this.resources.find(
                (relatedResource) => relatedResource.data.name === field.name
            )

            if (!relatedResource) {
                throw [
                    {
                        message: `The related resource ${field.name} was not found.`,
                    },
                ]
            }

            if (field.component === 'HasManyField') {
                // first we'll set all related belongs to values to null on the related resource
                const relatedBelongsToField = relatedResource.data.fields.find(
                    (field) =>
                        field.component === 'BelongsToField' &&
                        field.name === resource.data.name
                )

                if (!relatedBelongsToField) {
                    throw [
                        {
                            message: `A related BelongsTo relationship must be registered on the ${relatedResource.data.name} resource. This will link the ${resource.data.name} to the ${relatedResource.data.name} resource.`,
                        },
                    ]
                }

                // go to posts table, find all related posts and update the user_id field to be the null
                await this.db.updateManyWhere(
                    relatedResource,
                    { [relatedBelongsToField.databaseField]: modelId },
                    { [relatedBelongsToField.databaseField]: null }
                )

                // finally, go to posts table, find all related posts (new ones from request body) and update the user_id field to be modelId
                await this.db.updateManyByIds(
                    relatedResource,
                    relationshipFieldsPayload[field.inputName],
                    {
                        [relatedBelongsToField.databaseField]: modelId,
                    }
                )
            }
        }
    }

    public async createRelationalFields(
        resource: ResourceContract,
        payload: DataPayload,
        model: any
    ) {
        const {
            relationshipFieldsPayload,
        } = this.breakFieldsIntoRelationshipsAndNonRelationships(
            this.getResourceFieldsFromPayload(payload, resource),
            resource
        )

        const relationshipFields = resource
            .serialize()
            .fields.filter((field) => field.isRelationshipField)

        for (let index = 0; index < relationshipFields.length; index++) {
            const field = relationshipFields[index]
            const relatedResource = this.resources.find(
                (relatedResource) => relatedResource.data.name === field.name
            )

            if (!relatedResource) {
                throw [
                    {
                        message: `The related resource ${field.name} was not found.`,
                    },
                ]
            }

            if (field.component === 'HasManyField') {
                const relatedBelongsToField = relatedResource.data.fields.find(
                    (field) =>
                        field.component === 'BelongsToField' &&
                        field.name === resource.data.name
                )

                if (!relatedBelongsToField) {
                    throw [
                        {
                            message: `A related BelongsTo relationship must be registered on the ${relatedResource.data.name} resource. This will link the ${resource.data.name} to the ${relatedResource.data.name} resource.`,
                        },
                    ]
                }
                const valuesToUpdate: DataPayload = {
                    [relatedBelongsToField.databaseField]: model.id,
                }

                // go to posts table, find all related posts and update the user_id field to be the model.id
                this.db.updateManyByIds(
                    relatedResource,
                    relationshipFieldsPayload[field.inputName],
                    valuesToUpdate
                )
            }
        }
    }

    public async validateRequestQuery(
        {
            per_page: perPage,
            page,
            fields,
            search,
            filter,
            with: withRelationships,
            no_pagination: noPagination = 'false',
        }: Request['query'],
        resource: ResourceContract
    ) {
        let filters: FetchAllRequestQuery['filters'] = []

        if (typeof filter === 'object') {
            Object.keys(filter || {}).forEach((filterKey) => {
                const [field, operator] = filterKey.split(':')

                filters.push({
                    field,
                    operator: (operator as any) || 'equals',
                    value: ((filter || {}) as any)[filterKey],
                })
            })
        }

        const supportedOperators = [
            'equals',
            'contains',
            'not_equals',
            'exists',
            'doesnt_exist',
            'null',
            'not_null',
            'gt',
            'gte',
            'lt',
            'lte',
            'matches',
        ]

        const validFields = resource
            .serialize()
            .fields.filter((field) => !field.isRelationshipField)
            .map((field) => field.databaseField)

        const relationshipFields = resource
            .serialize()
            .fields.filter((field) => field.isRelationshipField)
            .map((field) => field.inputName)

        const parsedQuery = await validateAll(
            {
                perPage,
                page,
                search,
                filters,
                noPagination,
                fields: (fields as string)?.split(','),
                withRelationships: (withRelationships as string)?.split(','),
            },
            {
                perPage: 'number',
                page: 'number',
                fields: 'array',
                search: 'string',
                withRelationships: 'array',
                noPagination: 'string,in:true,false',
                'fields.*': 'in:' + validFields.join(','),
                'withRelationships.*': 'in:' + relationshipFields.join(','),
                filters: 'array',
                'filters.*.field':
                    'required|string|in:' + validFields.join(','),
                'filters.*.value': 'required|string',
                'filters.*.operator':
                    'required|string|in:' + supportedOperators.join(','),
            }
        )

        return parsedQuery
    }

    public async findAll(
        request: Request,
        resourceSlugOrResource: string | ResourceContract
    ) {
        const resource = this.findResource(resourceSlugOrResource)

        const {
            perPage,
            page,
            fields,
            search,
            filters,
            noPagination,
            withRelationships,
        } = await this.validateRequestQuery(request.query, resource)

        return this.db.findAll(resource, {
            perPage: perPage || resource.data.perPageOptions[0] || 10,
            page: page || 1,
            fields,
            search,
            filters,
            noPagination,
            withRelationships,
        })
    }

    public async findAllRelatedResource(
        request: Request,
        resourceId: string | number,
        resourceSlugOrResource: string | ResourceContract,
        relatedResourceSlugOrResource: string | ResourceContract
    ) {
        const resource = this.findResource(resourceSlugOrResource)
        const relatedResource = this.findResource(relatedResourceSlugOrResource)

        const relationField = resource.data.fields.find(
            (field) => field.name === relatedResource.data.name
        )

        const {
            perPage = 10,
            page = 1,
            filters,
            ...rest
        } = await this.validateRequestQuery(request.query, relatedResource)

        if (!relationField) {
            throw {
                status: 404,
                message: `Related field not found between ${resource.data.name} and ${relatedResource.data.name}.`,
            }
        }

        if (relationField.component === 'BelongsToManyField') {
            return this.db.findAllBelongingToMany(
                resource,
                relatedResource,
                resourceId,
                {
                    ...rest,
                    perPage,
                    page,
                    filters,
                }
            )
        }

        if (relationField.component === 'HasManyField') {
            const belongsToField = relatedResource.data.fields.find(
                (field) => field.name === resource.data.name
            )

            if (!belongsToField) {
                throw {
                    status: 404,
                    message: `Related 'belongs to' field not found between ${resource.data.name} and ${relatedResource.data.name}.`,
                }
            }

            return this.db.findAll(relatedResource, {
                ...rest,
                perPage,
                page,
                filters: [
                    ...filters,
                    {
                        field: belongsToField.databaseField,
                        value: resourceId,
                        operator: 'equals',
                    },
                ],
            })
        }

        return {}
    }

    public async findOneById(
        request: Request,
        resourceSlugOrResource: string | ResourceContract,
        id: number | string,
        withRelated?: string[]
    ) {
        const resource = this.findResource(resourceSlugOrResource)

        const { fields, withRelationships } = await this.validateRequestQuery(
            request.query,
            resource
        )

        const model = await this.db.findOneById(
            resource,
            id,
            fields,
            withRelated ? withRelated : withRelationships
        )

        if (!model) {
            throw {
                message: `Could not find a resource with id ${id}`,
                status: 404,
            }
        }

        return model
    }

    getValidationRules = (resource: ResourceContract, creationRules = true) => {
        const fields = resource.data.fields.filter((field) =>
            creationRules
                ? field.showHideField.showOnCreation
                : field.showHideField.showOnUpdate
        )

        const rules: {
            [key: string]: string
        } = {}

        fields.forEach((field) => {
            const serializedField = field.serialize()

            const fieldValidationRules = Array.from(
                new Set([
                    ...serializedField.rules,
                    ...serializedField[
                        creationRules ? 'creationRules' : 'updateRules'
                    ],
                ])
            ).join('|')

            if (fieldValidationRules) {
                rules[serializedField.inputName] = fieldValidationRules
            }
        })

        return rules
    }

    getResourceFieldsFromPayload = (
        payload: DataPayload,
        resource: ResourceContract
    ) => {
        let validPayload: DataPayload = {}

        resource.data.fields.forEach((field) => {
            const serializedField = field.serialize()

            if (Object.keys(payload).includes(serializedField.inputName)) {
                validPayload[serializedField.inputName] =
                    payload[serializedField.inputName]
            }
        })

        return validPayload
    }

    breakFieldsIntoRelationshipsAndNonRelationships = (
        payload: DataPayload,
        resource: ResourceContract
    ) => {
        const relationshipFieldsPayload: DataPayload = {}
        const nonRelationshipFieldsPayload: DataPayload = {}

        resource.data.fields.forEach((field) => {
            const serializedField = field.serialize()

            if (Object.keys(payload).includes(serializedField.inputName)) {
                if (serializedField.isRelationshipField) {
                    relationshipFieldsPayload[serializedField.inputName] =
                        payload[serializedField.inputName]
                } else {
                    nonRelationshipFieldsPayload[serializedField.inputName] =
                        payload[serializedField.inputName]
                }
            }
        })

        return {
            relationshipFieldsPayload,
            nonRelationshipFieldsPayload,
        }
    }

    validate = async (
        payload: DataPayload,
        resource: ResourceContract,
        creationRules: boolean = true,
        modelId?: string | number
    ): Promise<DataPayload> => {
        const {
            relationshipFieldsPayload,
            nonRelationshipFieldsPayload,
        } = this.breakFieldsIntoRelationshipsAndNonRelationships(
            this.getResourceFieldsFromPayload(payload, resource),
            resource
        )

        const parsedPayload: DataPayload = await validateAll(
            nonRelationshipFieldsPayload,
            this.getValidationRules(resource, creationRules),
            resource.data.validationMessages
        )

        await this.validateRelationshipFields(
            relationshipFieldsPayload,
            resource
        )

        await this.validateUniqueFields(
            nonRelationshipFieldsPayload,
            resource,
            creationRules,
            modelId
        )

        // then we need to validate all unique fields.
        return {
            parsedPayload,
            relationshipFieldsPayload,
        }
    }

    validateUniqueFields = async (
        payload: DataPayload,
        resource: ResourceContract,
        creationRules = true,
        modelId?: string | number
    ) => {
        const uniqueFields = resource
            .serialize()
            .fields.filter(
                (field) => field.isUnique && !field.isRelationshipField
            )

        for (let index = 0; index < uniqueFields.length; index++) {
            const field = uniqueFields[index]

            let exists: null | {} = null

            if (!payload[field.inputName]) {
                return
            }

            if (creationRules) {
                exists = await this.db.findOneByField(
                    resource,
                    field.databaseField,
                    payload[field.inputName],
                    ['id']
                )
            } else {
                exists = await this.db.findOneByFieldExcludingOne(
                    resource,
                    field.databaseField,
                    payload[field.inputName],
                    modelId!,
                    ['id']
                )
            }

            if (exists) {
                throw [
                    {
                        message: `A ${resource.data.name.toLowerCase()} already exists with ${
                            field.inputName
                        } ${payload[field.inputName]}.`,
                        field: field.inputName,
                    },
                ]
            }
        }
    }

    validateRelationshipFields = async (
        payload: DataPayload,
        resource: ResourceContract
    ) => {
        const fields = resource.data.fields
            .map((field) => field.serialize())
            .filter((field) => field.isRelationshipField)

        for (let index = 0; index < fields.length; index++) {
            const field = fields[index]
            const relatedResource = this.resources.find(
                (relatedResource) => relatedResource.data.name === field.name
            )

            if (!relatedResource) {
                throw [
                    {
                        message: `The related resource ${field.name} could not be found.`,
                        field: field.inputName,
                    },
                ]
            }

            if (!payload[field.inputName]) {
                return
            }

            if (field.component === 'HasManyField') {
                const allRelatedRows = await this.db.findAllByIds(
                    relatedResource,
                    payload[field.inputName],
                    ['id']
                )

                if (allRelatedRows.length !== payload[field.inputName].length) {
                    throw [
                        {
                            message: `Invalid values were provided for the related resource. Make sure all values provided exist in the database table ${relatedResource.data.table}`,
                            field: field.inputName,
                        },
                    ]
                }
            }

            if (field.component === 'BelongsToField') {
                if (
                    !(await this.db.findOneById(
                        relatedResource,
                        payload[field.inputName],
                        ['id']
                    ))
                ) {
                    throw []
                }
            }
        }
    }

    runAction = async (
        request: Request,
        resourceSlug: string,
        actionSlug: string
    ): Promise<ActionResponse> => {
        const resource = this.findResource(resourceSlug)

        const action = resource.data.actions.find(
            (action) => action.data.slug === actionSlug
        )

        if (!action) {
            throw {
                message: `Action ${actionSlug} is not defined on ${resourceSlug} resource.`,
                status: 404,
            }
        }

        const models = await this.db.findAllByIds(
            resource,
            request.body.models || []
        )

        const actionResource = createResourceFn(action.name).fields(
            action.data.fields
        )

        const { parsedPayload: payload } = await this.validate(
            request.body.form || {},
            actionResource
        )

        const response = await action.data.handler({
            request,
            models,
            payload,
            html: (html, status = 200) => ({
                html,
                type: 'html',
                status,
            }),
            errors: (errors, status = 200) => ({
                type: 'validation-errors',
                errors,
                status,
            }),
            notification: (notification, status = 200) => ({
                ...notification,
                type: 'notification',
                status,
            }),
            push: (route, status = 200) => ({
                type: 'push',
                status,
                route,
            }),
        })

        return response
    }

    getAdministratorById = (id: number | string) => {
        return this.db.getAdministratorById(id)
    }

    findUserByEmail = (email: string) => {
        return this.db.findUserByEmail(email)
    }

    getAdministratorsCount = () => this.db.getAdministratorsCount()
}
