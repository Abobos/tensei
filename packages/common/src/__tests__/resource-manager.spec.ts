import { Manager } from '../resources/Manager'

import db from './helpers/db'
import { Tag, Comment, User, Post } from './helpers/resources'

const setup = (resources = []) =>
    new Manager({} as any, [Tag, Comment, User, Post, ...resources], db as any)

describe('Manager', () => {
    describe('findResource', () => {
        test('can find a resource by string', () => {
            const resource = setup().findResource('users')

            expect(resource.data.name).toBe(User.data.name)
        })

        test('returns the resourceOrSlug if it is an instance', () => {
            const resource = setup().findResource(User)

            expect(resource.data.name).toBe(User.data.name)
        })

        test('throws an error if the resource is not found', () => {
            expect(() => setup().findResource('unknown-resource')).toThrow({
                message: 'Resource unknown-resource not found.',
            } as any)
        })

        test('throws an error if the slug is not passed in', () => {
            expect(() => setup().findResource(undefined as any)).toThrow({
                message: 'Resource undefined not found.',
            } as any)
        })
    })

    describe('create', () => {
        ;['posts', 'comments', 'users', 'tags'].forEach((slug) => {
            test(`validates payload before creating any record (${slug})`, async () => {
                db.create.mockClear()
                expect.assertions(2)

                try {
                    await setup().create({} as any)
                } catch (error) {
                    expect(error).toMatchSnapshot()
                }

                expect(db.create).not.toHaveBeenCalled()
            })
        })
    })

    describe('breakFieldsIntoRelationshipsAndNonRelationships', () => {
        test('correctly breaks payload into relationship fields and non relationships fields', () => {
            const {
                relationshipFieldsPayload,
                nonRelationshipFieldsPayload,
            } = setup().breakFieldsIntoRelationshipsAndNonRelationships(
                {
                    tags: [1, 2],
                    title: 'TEST_TITLE',
                    description: 'TEST_DESCRIPTION',
                },
                Post
            )

            expect(relationshipFieldsPayload).toEqual({
                tags: [1, 2],
            })

            expect(nonRelationshipFieldsPayload).toEqual({
                title: 'TEST_TITLE',
                description: 'TEST_DESCRIPTION',
            })
        })

        test('ignores fields not in payload', () => {
            const {
                relationshipFieldsPayload,
                nonRelationshipFieldsPayload,
            } = setup().breakFieldsIntoRelationshipsAndNonRelationships(
                {
                    tags: [1, 2],
                    title: 'TEST_TITLE',
                    description: 'TEST_DESCRIPTION',
                },
                Post
            )

            expect(
                Object.keys({
                    ...relationshipFieldsPayload,
                    ...nonRelationshipFieldsPayload,
                }).includes('content')
            ).toBeFalsy()
        })
    })

    describe('deleteById', () => {
        test('calls deleteById method from the db', () => {
            setup().deleteById('1')

            expect(db.deleteById).toHaveBeenCalled()
        })
    })
})
