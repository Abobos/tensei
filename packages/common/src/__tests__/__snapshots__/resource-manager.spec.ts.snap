// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Manager create validates payload before creating any record (comments) 1`] = `
Array [
  Object {
    "field": "title",
    "message": "The title is required.",
    "validation": "required",
  },
  Object {
    "field": "body",
    "message": "The body is required.",
    "validation": "required",
  },
]
`;

exports[`Manager create validates payload before creating any record (posts) 1`] = `
Array [
  Object {
    "field": "title",
    "message": "The title is required.",
    "validation": "required",
  },
  Object {
    "field": "description",
    "message": "The description is required.",
    "validation": "required",
  },
  Object {
    "field": "content",
    "message": "The content is required.",
    "validation": "required",
  },
  Object {
    "field": "av_cpc",
    "message": "The av_cpc is required.",
    "validation": "required",
  },
  Object {
    "field": "category",
    "message": "The category is required.",
    "validation": "required",
  },
  Object {
    "field": "user_id",
    "message": "The user_id is required.",
    "validation": "required",
  },
  Object {
    "field": "published_at",
    "message": "The published_at is required.",
    "validation": "required",
  },
  Object {
    "field": "scheduled_for",
    "message": "The scheduled_for is required.",
    "validation": "required",
  },
]
`;

exports[`Manager create validates payload before creating any record (tags) 1`] = `
Array [
  Object {
    "field": "name",
    "message": "The name is required.",
    "validation": "required",
  },
]
`;

exports[`Manager create validates payload before creating any record (users) 1`] = `
Array [
  Object {
    "field": "full_name",
    "message": "The full_name is required.",
    "validation": "required",
  },
  Object {
    "field": "email",
    "message": "The email is required.",
    "validation": "required",
  },
  Object {
    "field": "password",
    "message": "The password is required.",
    "validation": "required",
  },
]
`;
